python/pyiceberg/io/s3fs.py

```diff
@@ -0,0 +1,235 @@
+
+    def __len__(self) -> int:
+        """Returns the total length of the file, in bytes"""
+        object_info = self._s3.info(self.location)
+        if object_info.get("Size"):  # s3fs versions seem inconsistent on the case used for size
+            return object_info["Size"]
+        elif object_info.get("size"):
+            return object_info["size"]
+        raise RuntimeError(f"Cannot retrieve object info: {self.location}")
+
+    def exists(self) -> bool:
+        """Checks whether the location exists"""
+        return self._s3.lexists(self.location)
+
+    def open(self) -> S3fsInputStream:
+        """Create an S3fsInputStream for reading the contents of the file
+
+        Returns:
+            S3fsInputStream: An S3fsInputStream instance for the file located at `self.location`
+        """
+        return S3fsInputStream(s3_object=self._s3.open(self.location, "rb"))
+
+
+class S3fsOutputFile(OutputFile):
+    """An output file implementation for the S3fsFileIO
+
+    Args:
+        location(str): An S3 URI
+
+    Attributes:
+        location(str): An S3 URI
+    """
+
+    def __init__(self, location: str, s3):
+        self._s3 = s3
+        super().__init__(location=location)
+
+    def __len__(self) -> int:
+        """Returns the total length of the file, in bytes"""
+        object_info = self._s3.info(self.location)
+        if object_info.get("Size"):  # s3fs versions seem inconsistent on the case used for size
+            return object_info["Size"]
+        elif object_info.get("size"):
+            return object_info["size"]
+        raise RuntimeError(f"Cannot retrieve object info: {self.location}")
+
+    def exists(self) -> bool:
+        """Checks whether the location exists"""
+        return self._s3.lexists(self.location)
+
+    def create(self, overwrite: bool = False) -> S3fsOutputStream:
+        """Create an S3fsOutputStream for reading the contents of the file
+
+        Args:
+            overwrite(bool): Whether to overwrite the file if it already exists
+
+        Returns:
+            S3fsOutputStream: A wrapper for writing to an s3 object located at self.location
+
+        Raises:
+            FileExistsError: If the file already exists at `self.location` and `overwrite` is False
+
+        Note:
+            If overwrite is set to False, a check is first performed to verify that the file does not exist.
+            This is not thread-safe and a possibility does exist that the file can be created by a concurrent
+            process after the existence check yet before the output stream is created. In such a case, the default
+            behavior will truncate the contents of the existing file when opening the output stream.
+        """
+        if not overwrite and self.exists():
+            raise FileExistsError(f"Cannot create file, file already exists: {self.location}")
+        self._s3.touch(self.location)
+        return S3fsOutputStream(s3_object=self._s3.open(self.location, "wb"))
```

Comment:
What is thrown if this is called when the file already exists? There's a race condition if we check `exists`, so it would be great if we could just rely on an exception and re-raise the right one.